librz/main/rz-diff.c

```diff
@@ -1395,6 +1395,659 @@ static bool rz_diff_unified_files(DiffContext *ctx) {
+		free(opcodes_a);
+	}
+
+	// then from the match function
+	rz_list_foreach (result->unmatch_b, iter, bbi_b) {
+		opcodes_b = basic_block_opcodes(core_b, bbi_b);
+		if (!opcodes_b) {
+			break;
+		}
+
+		pj_o(pj); // { -- array object begin
+
+		pj_ko(pj, "pair"); // "pair": {
+		pj_knull(pj, "source");
+		graph_basic_block_json("match", bbi_b, pj);
+		pj_end(pj); // } -- "pair" end
+
+		pj_ko(pj, "similarity"); // "similarity": {
+		pj_ks(pj, "type", RZ_ANALYSIS_SIMILARITY_UNLIKE_STR);
+		pj_kd(pj, "score", 0.0);
+		pj_end(pj); // } -- "similarity" end
+
+		pj_ks(pj, "opcodes", opcodes_b);
+		pj_end(pj); // } -- array object end
+		free(opcodes_b);
+	}
+
+	pj_end(pj); // } -- result object end
+}
+
+static void diff_graph_function_detail_as_json(const char *object_name, RzAnalysisFunction *fcn, PJ *pj) {
+	pj_ko(pj, object_name); // "<object_name>: {" -- object begin
+
+	char *fcn_name = rz_str_escape_utf8_for_json(fcn->name, -1);
+	pj_ks(pj, "name", rz_str_get_null(fcn_name));
+	free(fcn_name);
+	pj_kn(pj, "offset", fcn->addr);
+	pj_ki(pj, "ninstr", fcn->ninstr);
+	pj_kn(pj, "nargs", rz_analysis_var_count_total(fcn, RZ_ANALYSIS_VAR_TYPE_ARGUMENT));
+	pj_kn(pj, "nlocals", rz_analysis_var_count_total(fcn, RZ_ANALYSIS_VAR_TYPE_LOCAL));
+	pj_kn(pj, "size", rz_analysis_function_linear_size(fcn));
+	pj_ki(pj, "stack", fcn->maxstack);
+	pj_ks(pj, "type", rz_analysis_fcntype_tostring(fcn->type));
+
+	pj_end(pj); // } -- <object_name> object end
+}
+
+static void diff_graph_as_json(RzCore *core_a, RzAnalysisFunction *fcn_a, RzCore *core_b, RzAnalysisFunction *fcn_b, RzAnalysisMatchResult *result, PJ *pj) {
+	pj_o(pj);
+	diff_graph_function_detail_as_json("source", fcn_a, pj);
+	diff_graph_function_detail_as_json("match", fcn_a, pj);
+	diff_graph_result_as_json(core_a, fcn_a, core_b, fcn_b, result, pj);
+	pj_end(pj);
+}
+
+static RzAnalysisFunction *find_best_matching_function(RzAnalysis *analysis_a, RzAnalysis *analysis_b, RzAnalysisFunction *find) {
+	RzAnalysisMatchPair *pair = NULL;
+	RzAnalysisFunction *match = NULL;
+	RzAnalysisMatchResult *result = NULL;
+	RzList *list_a = rz_list_new();
+	if (!list_a || !rz_list_append(list_a, find)) {
+		RZ_LOG_ERROR("rz-diff: cannot allocate and initialize RzList for function search\n");
+		goto fail;
+	}
+
+	result = rz_analysis_match_functions_2(analysis_a, list_a, analysis_b, analysis_b->fcns);
+	if (result && rz_list_length(result->matches) > 0) {
+		pair = (RzAnalysisMatchPair *)rz_list_first(result->matches);
+		match = (RzAnalysisFunction *)pair->pair_b;
+	}
+
+fail:
+	rz_analysis_match_result_free(result);
+	rz_list_free(list_a);
+	return match;
+}
+
+/**
+ * \brief Generate a json or graphviz dot output of the graph and its data.
+ *
+ * Each node that doesn't match 100% with the other function will include
+ * a unified diff of the assembly of the same basic block.
+ * */
+RZ_API bool core_show_function_diff(RzCore *core_a, ut64 addr_a, RzCore *core_b, ut64 addr_b, DiffMode mode) {
```

Comment:
Why API? It shouldn't be exported, IMHO.

Reply:
i forgot to remove it, should be static.